### PR TITLE
tests: fix kbs host name and port resolution

### PR DESCRIPTION
**Pull request by ryansavino**
Removed conditional block to get kbs host name and port number via host IP and node port. Instead force grab from the service directly.

Improved logic in functions.

Fixes: #10549

### CODE REVIEW

**Comment by ryansavino** [comment]
The removed block of code from the two methods detects the hostname from the node and node port instead of the service.

This is not working on our baremetal deployments. Could also be caused by the service not being exposed correctly to the host. @fidencio @stevenhorsman @wainersm @BbolroC

**Review by fidencio — CHANGES_REQUESTED** [review]
This breaks TDX.
```sh
  /home/sdp/actions-runner/_work/kata-containers/kata-containers/tests/integration/kubernetes/confidential_kbs.sh: line 428: host: unbound variable
  Trying to connect at http://:8080. Timeout=350
  ERROR: service seems to not respond on http://:8080 host
  curl: (3) URL rejected: No host part in the URL
```

**Comment by fidencio** [comment]
This also breaks `qemu-coco-dev` on zvsi:
```sh
  /home/ansible/gha-runner/_layout/_work/kata-containers/kata-containers/tests/integration/kubernetes/confidential_kbs.sh: line 428: host: unbound variable
  Trying to connect at http://:8080. Timeout=350
  ERROR: service seems to not respond on http://:8080 host
  curl: (3) URL using bad/illegal format or missing URL
```

**Comment by ryansavino** [comment]
Are you using ingress? I may need some help here from various env owners to test this, because the current method is not working for snp.

Is the service accessible for these deployments? Not sure if it's setup differently. With the kbs deployed, can you run:
```
KBS_SVC_NAME="kbs"
KBS_NS="coco-tenant"
kubectl get svc "$KBS_SVC_NAME" -n "$KBS_NS" -o jsonpath='{.spec.clusterIP}'
```

**Comment by fidencio** [comment]
> Are you using ingress? I may need some help here from various env owners to test this, because the current method is not working for snp.

Both TDX and s390x are using nodeport as the ingress type, as we do rely on a KBS "kustomization" that relies on `nodeport`.

@BbolroC will be able to help, as I'm going for vacation and coming back on November 25th.

**Comment by fidencio** [comment]
Looking at git history, the last relevant change on what you're touching was done by 6b0eaca4d467902073f9e356e732589d300a81bc ... and that was committed in at `Tue Mar 26 16:59:00 2024`.

AMD CI was up and running for several months after that, which makes me think that the code path you're touching is not actually what you're looking for.

File: tests/integration/kubernetes/confidential_kbs.sh [hunk 1]
```diff
@@ -418,32 +418,44 @@ function kbs_k8s_deploy() {
 # otherwise the cluster IP.
 #
 kbs_k8s_svc_host() {
+	local host
+
+	# Ingress
```
Comment (BbolroC) [inline tests/integration/kubernetes/confidential_kbs.sh:423]: For s390x, I got 

```
$ KBS_SVC_NAME="kbs"
$ KBS_NS="coco-tenant"
$ kubectl get svc "$KBS_SVC_NAME" -n "$KBS_NS" -o jsonpath='{.spec.clusterIP}'
10.43.140.148
```

Based on your logic, I've confirmed that adding:

```
host=""
```

here resolves the CI issue. Thanks.

Comment (ryansavino) [reply tests/integration/kubernetes/confidential_kbs.sh:423]: My bad, I've fixed the unbound variable like you mentioned here. Tests running.

**Comment by ryansavino** [comment]
> Looking at the git history, the last relevant change on what you're touching was done by [6b0eaca](https://github.com/kata-containers/kata-containers/commit/6b0eaca4d467902073f9e356e732589d300a81bc) ... and that was committed at `Tue Mar 26 16:59:00 2024`.
> 
> Or [c36c300](https://github.com/kata-containers/kata-containers/commit/c36c300fd60f70026700942fcc4a0da36eff3948) ... and that was commited at `Wed Jun 12 20:01:59 2024`.
> 
> AMD CI was up and running for several months after those two got merged, which makes me think that the code path you're touching is not actually what you're looking for.

@fidencio I saw this history as well, however the KBS URL is not accessible by the existing logic on the SNP node. We validated that using the service IP and port instead of the host name and node port worked and the tests were passing. @AdithyaKrishnan is submitting a PR to re-enable the SNP CI and we can test further. I did mention that maybe something changed in the way that the service is exposed through the host. Maybe that's what I should be looking for instead, but initial look at the code, I didn't see anything service specific. Maybe I missed something.

**Comment by wainersm** [comment]
> > Looking at the git history, the last relevant change on what you're touching was done by [6b0eaca](https://github.com/kata-containers/kata-containers/commit/6b0eaca4d467902073f9e356e732589d300a81bc) ... and that was committed at `Tue Mar 26 16:59:00 2024`.
> > Or [c36c300](https://github.com/kata-containers/kata-containers/commit/c36c300fd60f70026700942fcc4a0da36eff3948) ... and that was commited at `Wed Jun 12 20:01:59 2024`.
> > AMD CI was up and running for several months after those two got merged, which makes me think that the code path you're touching is not actually what you're looking for.
> 
> @fidencio I saw this history as well, however the KBS URL is not accessible by the existing logic on the SNP node. We validated that using the service IP and port instead of the host name and node port worked and the tests were passing.

Yeah, I share same opinion of @fidencio , something changed on snp environment as that function seems stable for a long time. Anyway, I'm not against applying the fix but we should test it very well to not break other jobs.

Ah, I'd like to test this change doesn't break with kcli (which is not covered in any job but some people use to reproduce CI locally).

> @AdithyaKrishnan is submitting a PR to re-enable the SNP CI and we can test further. I did mention that maybe something changed in the way that the service is exposed through the host. Maybe that's what I should be looking for instead, but initial look at the code, I didn't see anything service specific. Maybe I missed something.

**Comment by fidencio** [comment]
I personally would prefer having this one merged after the release, just to make sure nothing else breaks and, if does, there is time to fix without the release pressure.

@kata-containers/architecture-committee

**Comment by ryansavino** [comment]
Just spoke with Wainer. Here's what we're going to do.

1. Merge Adi's PR #10558 to re-enable SNP CI, but skip confidential tests
2. Test this PR without the fix, but un-skipping the confidential tests to reproduce failure
3. Test fix in this PR to validate that it actually fixes the problem

**Comment by ryansavino** [comment]
Performing step 2 right now from last comment. Apparently, the node name and node port are working on this test run for snp. I've stripped out my commit/fix and just re-enabled the tests. So I think the issue may be around how the kbs is being deployed. This would reinforce the idea that the issue is intermittent and not related specifically to the commit from 5 months ago as @fidencio pointed out.
https://github.com/kata-containers/kata-containers/actions/runs/11938148085/job/33276645856?pr=10550

However, it is my opinion that the service ip and service port are the more correct values to use here as a service could technically be spread across multiple nodes in a cluster. Our test systems are typically one node cluster deployments. I know @wainersm had some concerns about some other implementations working. So let me know if you guys want me to re-add this commit/fix back in and get this merged.

Either way the snp ci is passing on this run. Can't really troubleshoot further until the error reproduces itself.

**Comment by ryansavino** [comment]
@fidencio the tdx job did fail with this test: k8s-job.bats

```
# Events:
#   Type     Reason     Age   From               Message
#   ----     ------     ----  ----               -------
#   Normal   Scheduled  59s   default-scheduler  Successfully assigned kata-containers-k8s-tests/job-pi-test-twm47 to 0007e9351d34
#   Normal   Pulling    43s   kubelet            Pulling image "quay.io/prometheus/busybox:latest"
#   Normal   Pulled     42s   kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 648ms (648ms including waiting). Image size: 1267792 bytes.
#   Normal   Created    42s   kubelet            Created container pi
#   Warning  Failed     35s   kubelet            Error: failed to start containerd task "pi": No such file or directory (os error 2)
```

**Comment by stevenhorsman** [comment]
> @fidencio the tdx job did fail with this test: k8s-job.bats
> 
> ```
> # Events:
> #   Type     Reason     Age   From               Message
> #   ----     ------     ----  ----               -------
> #   Normal   Scheduled  59s   default-scheduler  Successfully assigned kata-containers-k8s-tests/job-pi-test-twm47 to 0007e9351d34
> #   Normal   Pulling    43s   kubelet            Pulling image "quay.io/prometheus/busybox:latest"
> #   Normal   Pulled     42s   kubelet            Successfully pulled image "quay.io/prometheus/busybox:latest" in 648ms (648ms including waiting). Image size: 1267792 bytes.
> #   Normal   Created    42s   kubelet            Created container pi
> #   Warning  Failed     35s   kubelet            Error: failed to start containerd task "pi": No such file or directory (os error 2)
> ```

We saw that same failure on the non-tee coco nightly yesterday. My experience a couple of years back when porting that test to CAA was that it was very flakey and we had to attempt it multiple times and just check if one of them worked, so maybe we arefinally seeing that in kata as well?

**Review by fidencio — APPROVED** [review]
lgtm as it only reverts the changes to enable the tests for snp.

**Comment by ryansavino** [comment]
Okay, thanks @wainersm. I think I understand how this works now. I suppose the issue really lies with how the kbs is being deployed - if there is some intermittent issue there. I will wait to see if the error pops up again and look into any issues with the kbs deployment at that time.

Closing this PR.

**Pull request closed without merging**